### PR TITLE
Make it possible to use AMD and Intel GPUs as well as Apple Silicon (backend-only)

### LINKED ISSUE #72: Use PyTorch's `torch.accelerator` instead of  `torch.cuda` to provide support for non-NVIDIA-GPUs
Currently the `device_list.py` file resolves devices like the following:

```
import torch

gpu_list = []
for i in range(torch.cuda.device_count()):
    device_name = torch.cuda.get_device_properties(i).name
    gpu_list.append(device_name)

print(gpu_list)
```

As far as I can see the code doesn't rely on pure-CUDA features, so using the [accelerator](https://docs.pytorch.org/docs/stable/accelerator.html) api would mean being able use other accelerators automatically assuming the right version of pytorch is installed.

Currently, ROCm also gets enumerated by the CUDA-api but Intel GPUs do not. For Intel GPUs it might(?) be necessary to use [XPU variants](https://docs.pytorch.org/docs/stable/xpu.html) for things like memory stats if I read the code correctly, though the API seems the same. Maybe during initialization it would work to select the appropriate type based on GPU usage?

### PR DESCRIPTION
**Pull request by Lolle2000la**
Closes #72

I plan to eventually get a AMD graphics card and was thinking about using a Intel Arc GPU for my NAS, so tried to make the backend run on non-NVIDIA-GPUs.

Ironically, I only have access to NVIDIA GPUs (RTX 2070 Super and GTX 1070), but I can confirm that this works on those without regression. I have tested it for the past few days on my [manga library management app](https://github.com/Lolle2000la/MangaIngestWithUpscaling), where I include the backend from this repo. In it, I included my forked version to use these changes.

If you have access to a Intel or AMD GPU (or a Apple Silicon device, i.e. a MacBook, Mac Mini, etc.), please help me by testing these changes. For that you need to run install the right packages, preferably installed into a virtual environment:

#### AMD

```sh
pip install torch==2.7.1 torchvision==0.22.1 --extra-index-url https://download.pytorch.org/whl/rocm6.3 chainner_ext==0.3.10 numpy==2.2.5 opencv-python-headless==4.11.0.86 psutil==6.0.0 pynvml==11.5.3 pyvips==3.0.0 pyvips-binary==8.16.1 rarfile==4.2 sanic==24.6.0 spandrel_extra_arches==0.2.0 spandrel==0.4.1 --no-warn-script-location
```

#### Intel GPU

```sh
pip install torch==2.7.1 torchvision==0.22.1 --extra-index-url https://download.pytorch.org/whl/xpu chainner_ext==0.3.10 numpy==2.2.5 opencv-python-headless==4.11.0.86 psutil==6.0.0 pynvml==11.5.3 pyvips==3.0.0 pyvips-binary==8.16.1 rarfile==4.2 sanic==24.6.0 spandrel_extra_arches==0.2.0 spandrel==0.4.1 --no-warn-script-location
```

#### Apple Silicon (Mac)

```sh
# no extra index necessary
pip install torch==2.7.1 torchvision==0.22.1 chainner_ext==0.3.10 numpy==2.2.5 opencv-python-headless==4.11.0.86 psutil==6.0.0 pynvml==11.5.3 pyvips==3.0.0 pyvips-binary==8.16.1 rarfile==4.2 sanic==24.6.0 spandrel_extra_arches==0.2.0 spandrel==0.4.1 --no-warn-script-location
```

### CODE REVIEW

**Comment by BlueD** [comment]
There are no rocm for windows yet. So I don't it's possible for amd gpu now.

**Comment by Lolle2000la** [comment]
> Sorry about the delay, this is great, thanks.

No problem, I'm sure you all have a life as well :)

> Before merging I'd like to get the backend piece automated as well, so end users wouldn't have to worry about manually installing the correct dependencies for their system.

Do you mean the installation of the backend within the GUI or the installation following the instructions? 

For both of those, if older CUDA/ROCm versions aren't a concern, the same [pyproject.toml](https://github.com/the-database/MangaJaNaiConverterGui/blob/main/MangaJaNaiConverterGui/backend/src/pyproject.toml) can be used, with only the 
`extra-index-url = ["https://download.pytorch.org/whl/cu121"]`-part modified. I'm not sure what would be the most elegant way for that though.

In my variant I also used `opencv-python-headless` since there are a number of annoying circumstances where `opencv-python` can fail depending on the host.

> I also don't have access to non-NVIDIA systems, so we'll need to rely on the community for testing.

I thought about renting a AMD GPU instance on vast.ai, but nobody is hosting one of those :sweat_smile: 

It would maybe be a good idea to put an experimental-label onto that then.

**Review by the-database — APPROVED** [review]
